userspace: Introduce test in svsm

n-ramacciotti opened the pull request:
The PR enables running tests in svsm userspace by:
- Adding a custom test runner in `userlib` via the `test_runner` feature, which ~~is included in `userinit`~~ can be included in userspace binaries ~~(like in `userinit` in the `[testing]` commits)~~ for the `x86_64-unknown-none` target during tests.
- The `test-runner` features enables `test` crate in `userlib` only for `x86_64-unknown-none` target. For other targets, the feature is included but the optional `test` dependency is ignored.
- ~~Updating the ~~qemu~~ test configurations to build userspace binaries via Makefile  with the `LINK_TEST` environment variable.~~
- ~~user binaries are compiled in a secondary `makefile` in the `user` folder. These rules are called by the root makefile and can have `FEATURES_TEST` AND `USERMODULE_FEATURES_TEST` customizable (see https://github.com/coconut-svsm/svsm/pull/926#issuecomment-4712818365)~~
- ~~User-space binaries can be compiled via `Makefile`, updating the tests configurations to include `test_in_svsm` configuration with `LINK_TEST` (like in `[testing]` commits)~~
- ~~Moving filesystem initialization after kernel tests to ensure an empty filesystem for tests that require it.~~
- Changing some filesystem tests to account for an already populated filesystem.
- Catching test failures in userspace via the exit syscall.

## 
~~This PR is marked as draft as it depends on the changes from:~~
- ~~[#909](https://github.com/coconut-svsm/svsm/pull/909): this simplify test creation with xbuild~~

These changes were originally included in [#913](https://github.com/coconut-svsm/svsm/pull/913), which also introduced a userspace allocator, but as suggested, I split them into multiple PRs.

All feedback and suggestions are welcome

n-ramacciotti committed:
user/init: Add the test in svsm compilation

Update the test configurations to build `init`
using the Makefile.

Add the options in `build.rs` for both normal and
test use. In particular, add the `test_in_svsm`
flag.

Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

n-ramacciotti committed:
kernel/filesystem: Allow non empty root during tests

Some tests, such as `create_dir()` in `kernel/src/fs/filesystem.rs`,
require an empty filesystem. If the filesystem is already populated,
those tests will fail. Until now, the filesystem has always been
empty during tests, but this is no longer the case. Change the
assertions to check that the root contains the desired entry instead
of requiring an exact match. This change only affects checks on the
root directory, not on subdirectories created during tests.

Suggested-by: Luigi Leonardi <leonardi@redhat.com>
Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

n-ramacciotti committed:
kernel/filesystem: Allow non empty root during tests

Some tests, such as `create_dir()` in `kernel/src/fs/filesystem.rs`,
require an empty filesystem. If the filesystem is already populated,
those tests will fail. Until now, the filesystem has always been
empty during tests, but this is no longer the case. Change the
assertions to check that the root contains the desired entry instead
of requiring an exact match. This change only affects checks on the
root directory, not on subdirectories created during tests.

Suggested-by: Luigi Leonardi <leonardi@redhat.com>
Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

n-ramacciotti committed:
user/init: Move tests into a dedicated test function

Currently, the init process prints a string and then performs
some checks on global mutable data.

With the new test runner, these tests can be moved into a dedicated
test function that can be executed during runtime testing.

Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

n-ramacciotti committed:
user/lib: Introduce custom test runner

Introduce a custom test runner in `userlib` for executing tests
in userspace, following the same pattern used in the kernel.
Enable it via the `test_runner` feature, which also brings the
custom `test` crate from the workspace only when testing with the
`x86_64-unknown-none` target.

Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

n-ramacciotti committed:
user/lib: Introduce custom test runner

Introduce a custom test runner in `userlib` for executing tests in
userspace, following the same pattern used in the kernel.

Enable it via the `test_runner` feature, which also brings the custom
`test` crate from the workspace only when testing with the
`x86_64-unknown-none` target. If the feature is enabled for a different
target, Cargo will ignore the custom `test` crate.

This is because there is a conflict with the `test` crate during normal
testing, resulting in the  custom crate being imported instead of the
standard one. This causes the following error when running Clippy:

```
$ cargo clippy --all-features --package userlib --tests
    Checking test v0.1.0 (...)
    ...
    Checking userlib v0.1.0 (...)
error[E0425]: cannot find function `test_main_static` in crate `test`
```

Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

n-ramacciotti committed:
kernel/test-in-svsm: Run userspace tests after kernel tests

After the kernel tests complete, instead of exiting, launch
the userspace binaries to run userspace tests, then exit.

When a userspace test fails, the panic handler calls `exit(!0)`.
Update `sys_exit()` to report test failures to QEMU and exit
from the tests in SVSM.

Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

n-ramacciotti committed:
kernel/test-in-svsm: Run userspace tests after kernel tests

Add a new function that wraps a loop which calls `exec` for each user
binary, compiled to run userspace tests.

After the kernel tests complete, instead of exiting, call the new
function and then exit.

For simplicity, user binaries are listed in `/test`. If the latter is
not found, it is assumed that there are no tests to run.

Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

stefano-garzarella reviewed (commented):
Thanks for this!

BTW do we need any changes in the CI to run these tests, or they are already covered?

luigix25 commented on kernel/src/svsm.rs:
```diff
@@ -472,6 +472,22 @@ pub unsafe fn svsm_main(li: usize) {
 fn test_in_svsm_task(_context: usize) {
     crate::test_main();
+
+    #[cfg(test_in_svsm)]
```
I'm not a test expert, but why is this needed?

n-ramacciotti commented on kernel/src/svsm.rs:
```diff
@@ -472,6 +472,22 @@ pub unsafe fn svsm_main(li: usize) {
+        use svsm::fs::opendir;
+        use svsm::task::exec_user;
+
+        match exec_user("/init", opendir("/").expect("Failed to find FS root")) {
```
In this configuration, user tests start after kernel tests. 
It is possible to move `exec_user` and `testing::exit()` out of `test_in_svsm_task` and to place them before the `#[cfg(not(test))]`, removing the duplication from that section

n-ramacciotti committed:
user/init: Introduce test entry point

Introduce a `lib.rs` test entry point in the `userinit`
crate. Enable custom test framework and import the new
custom test runner from `userlib` by  enabling the
`test_runner` feature during tests on the
`x86_64-unknown-none` target.

Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

n-ramacciotti commented on Makefile:
```diff
@@ -198,7 +198,8 @@ clippy:
 	cargo clippy ${CLIPPY_OPTIONS} --all-features --package stage1 --target x86_64-unknown-none -- ${CLIPPY_ARGS} ${STAGE1_RUSTC_ARGS}
-	cargo clippy ${CLIPPY_OPTIONS} --all-features --workspace --tests --exclude packit -- ${CLIPPY_ARGS}
+	cargo clippy ${CLIPPY_OPTIONS} --all-features --workspace --tests --exclude packit --exclude userlib -- ${CLIPPY_ARGS}
+	cargo clippy ${CLIPPY_OPTIONS} --package userlib --tests -- ${CLIPPY_ARGS}
```
Using clippy with `--tests` and `--all-features` creates a conflict on the test crate. Given that the `test_runner` feature is enabled, `test` from the workspace is imported instead of the standard one and the following error occurs:
```
$ cargo clippy --all-features --package userlib --tests
    Checking test v0.1.0 (...) 
    ...
    Checking userlib v0.1.0 (...)
error[E0425]: cannot find function `test_main_static` in crate `test`
```

I had to disable `--all-features` to avoid this issue.

stefano-garzarella commented on kernel/src/svsm.rs:
```diff
@@ -497,7 +497,20 @@ pub unsafe fn svsm_main(li: usize) {
+
     crate::test_main();
+
+    match exec_user("/init", opendir("/").expect("Failed to find FS root")) {
```
> > Why that test should be in init ?
> 
> The current behavior of `init`(see [e8d571c](https://github.com/coconut-svsm/svsm/commit/e8d571c18b06bcce144744fffcff50e28e20c483)) is to perform some checks that I moved into `test_memory_check`. I don't think this check will be necessary in the future, it was just to show the behavior of the PR.

I think we can remove them, or leave them as they are now or move in some other place, but IMO init's main should not have any `#[cfg(test)]`

n-ramacciotti committed:
[RFC] user/init: Introduce main test function

Introduce a different main function used only
when testing that calls the custom test
harness

Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

n-ramacciotti commented:
V7:
- Rebased on the latest main.
- Introduced second main in `init` for testing purposes only, which calls the custom test harness.
  This is to avoid changing the normal main.

n-ramacciotti commented on user/init/src/main.rs:
```diff
@@ -7,12 +7,20 @@ use userlib::*;
 
 declare_main!(main);
 
+#[cfg(not(test))]
```
~~This commit will be squashed. It has been split here only for ease of viewing.~~

Another possibility consists in moving the test harness directly inside `declare_main` macro:

```
macro_rules! declare_main {
 // ...
 #[cfg(not(test))]
 {
       let main_fn: fn() -> u32 = $path;
       let ret = main_fn();
       exit(ret);
 }
  #[cfg(test)]
  {
        crate::userspace_test_main();
       exit(0)
  }
```

However, this requires:
- `#[allow(clippy::crate_in_macro_def)]` when defining the macro
- `#[allow(dead_code)]` for main function, as it is not used when testing

n-ramacciotti commented:
v8:
 - Moved the optional dependency `test` in  `userlib` under the target `x86_64-unknown-none`. This resolves [this discussion](https://github.com/coconut-svsm/svsm/pull/926#discussion_r2847940734). The feature `test_runner` still enables the optional dependency, but only when compiling for `x86_64-unkwown-none`, otherwise cargo ignores it.

n-ramacciotti committed:
user/test-in-svsm: Compile binaries via Makefile

The user binaries need to be compiled in the same way
as the kernel for testing inside svsm. They require
`LINK_TEST` to enable the `test_in_svsm` configuration
flag, a custom runner to copy the binary to the bin
folder and the nightly toolchain for the
`x86_64-unknown-none` target. They also require
compilation with `RUSTDOC=TRUE` to address the same kernel
issue, even though they currently have no documentations
tests and would still pass without it.

As the number of user binaries can increase, including
them all, with the same pattern repeated multiple times,
would make the Makefile very large.

Add a static pattern rule to extract the module from
the complete target (`bin/test-module.elf`) and use
it in the new command.

Update `.PHONY` accordingly.

The runner copies the binary to the bin folder
and requires a different path for the user-space
(`../../bin/test-module.elf`) and for the kernel
(`../bin/test-kernel.elf`). Simplify the command by
introducing `MAKEFILE_DIR` to point to the root of
the project.

Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

n-ramacciotti commented:
v9:
- Added to `Makefile` a static pattern rule for user binaries instead of repeating the same command for each one of them

n-ramacciotti committed:
[RFC] makefile: Introduce new variable

When testing inside svsm, the new user modules will
require features to be specified.

There is no problem when using `xbuild`, where we can
specify the features in the `build recipes`, but
when compiling directly with `make bin/test-module.elf`
without specifying `FEATURES_TEST`, it will take the default
one specified in the makefile (like `vtpm`). However, these
features are related to the kernel.

One possible solution would be to avoid specifying default
features for `FEATURES_TEST`, but this would require
updating every test that depends on `SVSM_ARGS_TEST`
and doesn't specify the features, like `make test` in
the CI. It would also require updating the `build recipes`
when using `xbuild` to include all these features by default.
However, this would mean that they would always be included
even when I only want a subset of them.

Another solution would be to introduce `MODULE_ARGS_TEST`
which would only get the features if they are not defined in
the makefile. This could then be used for
`bin/test-kernel.elf` and all future `bin/test-module.elf`

Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

n-ramacciotti committed:
[RFC] kernel/test-in-svsm: Use static pattern rule

This can be squashed in the previous commit, where the
rule is added.

Now that the static pattern rule is present and that
kernel compilation is very similar to that of all user
modules during test in svsm, use the static pattern rule
also for the kernel.

Change the name from `bin/test-kernel.elf` to
`bin/test-svsm.elf` as the name of the package is
`svsm` and the rule extracts this.

Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

n-ramacciotti commented:
v10:
- rebased on latest main
- Moved `bin/test-kernel.elf` in the new static pattern rule. (and renamed to bin/test-svsm.elf)
- Changed default features assignment when directly launching `make bin/test-module.elf`

stefano-garzarella commented on user/lib/src/testing.rs:
```diff
@@ -0,0 +1,31 @@
+// SPDX-License-Identifier: MIT
+
+use crate::{console_print, print, println};
```
is `console_print` really used?

n-ramacciotti replied:
When I tried to compile without it, I got the following error message:
```
error[E0425]: cannot find function `console_print` in this scope
  --> user/lib/src/console.rs:38:27
   |
38 |         ($($arg:tt)*) => (console_print(format_args!($($arg)*)))
   |                           ^^^^^^^^^^^^^ not found in this scope
   |
  ::: user/lib/src/testing.rs:7:5
   |
7  |     println!("running {} user tests", test_cases.len());
   |     --------------------------------------------------- in this macro invocation
   |
   = note: this error originates in the macro `print` which comes from the expansion of the macro `println` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider importing this function through its public re-export
  --> user/lib/src/testing.rs:3:1
   |
3  + use crate::console_print;
   |

```

stefano-garzarella replied:
What about this?

```diff
diff --git a/user/lib/src/console.rs b/user/lib/src/console.rs
index a79c56ba..a04c2b2e 100644
--- a/user/lib/src/console.rs
+++ b/user/lib/src/console.rs
@@ -35,7 +35,7 @@ pub fn console_print(args: fmt::Arguments<'_>) {
 
 #[macro_export]
 macro_rules! print {
-        ($($arg:tt)*) => (console_print(format_args!($($arg)*)))
+        ($($arg:tt)*) => ($crate::console_print(format_args!($($arg)*)))
 }
 
 #[macro_export]
```

n-ramacciotti replied:
Right! I will add a new commit for this

stefano-garzarella commented on kernel/src/svsm.rs:
```diff
@@ -618,7 +618,35 @@ pub unsafe fn svsm_main(li: usize) {
+
     crate::test_main();
+
+    // What is a good way of identifying all
```
Please prefix these questions with something like TODO: or RFC:

stefano-garzarella commented on kernel/src/svsm.rs:
```diff
@@ -618,7 +618,35 @@ pub unsafe fn svsm_main(li: usize) {
+    // should be tested
+    let files = list_dir("").unwrap();
+    log::info!("Files: {:?}", files);
+    // Should I move this logic in a userspace
```
As we discussed, moving in userspace will require to add syscall to export wait_for_termination behavior, right?

n-ramacciotti replied:
Yes, it would require a new system call. For now, it can be maintained in the kernel. I will remove the comment

stefano-garzarella commented on kernel/src/svsm.rs:
```diff
@@ -618,7 +618,35 @@ pub unsafe fn svsm_main(li: usize) {
     crate::test_main();
+
+    // What is a good way of identifying all
+    // userspace binaries? What about
```
How do we create that file?
I'd like to keep it as simple as possible.

n-ramacciotti replied:
My idea was to modify xbuild by adding a new field to the recipe:
```
/// Components for the filesystem image.
#[derive(Debug, Clone, Deserialize)]
pub struct FsConfig {
    modules: HashMap<String, ComponentConfig>,
    ##
    files : ..
    ##
}
```

I should look into these more closely, though.

n-ramacciotti replied:
Does it make sense or values hardcoded in the code would be enough? Maybe this could be addressed in a future PR, using hardcoded values for now

00xc replied:
> My idea was to modify xbuild by adding a new field to the recipe:
> 
> ```
> /// Components for the filesystem image.
> #[derive(Debug, Clone, Deserialize)]
> pub struct FsConfig {
>     modules: HashMap<String, ComponentConfig>,
>     ##
>     files : ..
>     ##
> }
> ```
> 
> I should look into these more closely, though.

Could we just place test binaries in a dedicated path? We already treat `/init` specially, we could have `/test/*` with all test binaries.

n-ramacciotti replied:
Oh right! That's even better. Thanks for the suggestion!

stefano-garzarella commented on kernel/src/svsm.rs:
```diff
@@ -618,7 +618,35 @@ pub unsafe fn svsm_main(li: usize) {
+                wait_for_termination(task);
+            }
+            Err(e) => {
+                log::info!("Failed to launch {file}: {e:?}");
```
maybe error instead of info?

stefano-garzarella commented on kernel/src/svsm.rs:
```diff
@@ -618,7 +618,35 @@ pub unsafe fn svsm_main(li: usize) {
+    use svsm::fs::{list_dir, opendir};
+    use svsm::task::{exec_user, wait_for_termination};
+
     crate::test_main();
```
I spent quite some time to understand this.

What about renaming this in kernel_tests() and adding userspace_tests() with the code below?

n-ramacciotti replied:
> What about renaming this in kernel_tests() 

Sure, it makes much sense!

> adding userspace_tests() with the code below?

Do you mean another function available during testing which wraps the loop? it makes sense. I will add it

stefano-garzarella replied:
> > What about renaming this in kernel_tests()
> 
> Sure, it makes much sense!
> 
> > adding userspace_tests() with the code below?
> 
> Do you mean another function available during testing which wraps the loop? it makes sense. I will add it

yep

stefano-garzarella commented on kernel/src/syscall/class0.rs:
```diff
@@ -15,6 +15,15 @@ use core::ffi::c_char;
+        if exit_code != 0 {
+            use crate::testing::QEMUExitValue;
+            crate::testing::exit(QEMUExitValue::Fail);
+        }
```
yep, that would be much better IMO.

n-ramacciotti replied:
I'm working on saving the `exit_code`. It will be a separate PR

stefano-garzarella replied:
great, thanks!

stefano-garzarella commented on Makefile:
```diff
@@ -6,9 +6,13 @@ endif
 
 FEATURES_TEST ?= vtpm,virtio-drivers,block,vsock
 SVSM_ARGS_TEST += --no-default-features
+MODULE_ARGS_TEST += --no-default-features
```
I'm really confused by MODULE_ARGS_TEST and changes in the Makefile to build userspace modules. I think we should try to avoid this hack in some better way, but I need to think more for suggestions.

n-ramacciotti replied:
I agree that this is not the best solution.

For building userspace modules in a test environment, it is needed to follow the same pattern that is used for `make bin/test-kernel.elf`:

```makefile
bin/test-kernel.elf: bin
# RUSTDOC=true removes doctests, which is necessary as they do not work with
# custom test runners. See https://github.com/coconut-svsm/svsm/issues/705.
	RUSTDOC=true LINK_TEST=1 cargo +nightly test --package svsm \
	       ${CARGO_ARGS} \
	       ${SVSM_ARGS_TEST} \
		--target=x86_64-unknown-none \
		--config 'target.x86_64-unknown-none.runner=["sh", "-c", "cp $$0 ../$@"]'
```

It should be something like:
```makefile
bin/test-userinit.elf: bin
# RUSTDOC=true removes doctests, which is necessary as they do not work with
# custom test runners. See https://github.com/coconut-svsm/svsm/issues/705.
	RUSTDOC=true LINK_TEST=1 cargo +nightly test --package userinit \
	        ${CARGO_ARGS} \
	         ${SVSM_ARGS_TEST} \
		--target=x86_64-unknown-none \
		--config 'target.x86_64-unknown-none.runner=["sh", "-c", "cp $$0 ../../$@"]'
```

If I use `xbuild` (`make bin/coconut-qemu-test.igvm`) there is no problem as `xbuild` can sort the feature for module (svsm:attest, userinit:featureX) from a single `FEATURES=svsm:attest,userinit:featurex`.

The problem is there when using `make bin/test-userinit.elf` directly. As it would take `FEATURES_TEST` which contains features not related to `userinit` (vtpm, etc..).

This requires to specify a different `SVSM_ARGS_TEST`, lets say `USERINIT_ARGS_TEST`, but this would create a problem when using `make bin/coconut-test-qemu.igvm` as here `xbuild` should specify two different `FEATURES_SVSM` and `FEATURES_USERINIT` and this would expand as the number of usermodules increases.

Also all the default features should all be passed to `xbuild` like `SVSM_ARGS_TEST` now.

I added `MODULE_ARGS_TEST` which simply ignores `FEATURES_TEST` when calling `make bin/test-userinit.elf` directly.

A better solution would be to avoid to call back into the `makefile`, as @00xc suggested in #909:
`
bin/test-kernel.elf is still built from the Makefile for now, since it requires a special command line invocation, and moving that logic into xbuild would introduce too much complexity. Thus, we have xbuild call back into the Makefile to build it 
`

stefano-garzarella replied:
`SVSM_ARGS_TEST` contains args to build SVSM's tests, indeed, for example, for the `test` target, we use it only for `svsm` packages, so we should not use it in `bin/test-userinit.elf` target IMO:

```
test:
	cargo test ${CARGO_ARGS} ${SVSM_ARGS_TEST} --package svsm
	cargo test ${CARGO_ARGS} --workspace --exclude svsm
```

stefano-garzarella replied:
> A better solution would be to avoid to call back into the `makefile`, as @00xc suggested in #909:
> `bin/test-kernel.elf is still built from the Makefile for now, since it requires a special command line invocation, and moving that logic into xbuild would introduce too much complexity. Thus, we have xbuild call back into the Makefile to build it `

yep, agree on this, but if it's too complicated and the problem is just `FEATURES_TEST` default for a specific target, what about this:
```
bin/test-userinit.elf: FEATURES_TEST=
```

IMO we should keep target-svsm and target-user* separated.

n-ramacciotti replied:
> yep, agree on this, but if it's too complicated and the problem is just `FEATURES_TEST` default for a specific target, what about this:
> 
> ```
> bin/test-userinit.elf: FEATURES_TEST=
> ```

IIUC, this is a similar solution to the `MODULE_ARGS_TEST`, but in this case `FEATURE_TEST` would always be empty while with `MODULE_ARGS_TEST` you could still specify them (`FEATURES_TEST=.. make bin/test-userinit.elf`)


In this case `SVSM_ARGS_TEST` is used only for `make test` and it contains (by default) the same features as `make bin/test-kernel.elf`

n-ramacciotti replied:
> IMO we should keep target-svsm and target-user* separated.

We can separate them, but then we would have to specify a new variable for each one of them.
(only `USERINIT_ARGS_TEST` for now) and also `xbuild` should be updated accordingly (to set that value) as it will be possible to have different feature for each usermodule and a single `USERMODULE_ARGS_TEST` would not help


We should also concatenate all these variables into `--feature` of `xbuild`

stefano-garzarella replied:
> > yep, agree on this, but if it's too complicated and the problem is just `FEATURES_TEST` default for a specific target, what about this:
> > ```
> > bin/test-userinit.elf: FEATURES_TEST=
> > ```
> 
> IIUC, this is a similar solution to the `MODULE_ARGS_TEST`, but in this case `FEATURE_TEST` would always be empty while with `MODULE_ARGS_TEST` you could still specify them (`FEATURES_TEST=.. make bin/test-userinit.elf`)
> 
> In this case `SVSM_ARGS_TEST` is used only for `make test` and it contains (by default) the same features as `make bin/test-kernel.elf`

Sorry, I didn't get.

Did you try it?

You can still specify `FEATURE_TEST=foo make` and `FEATURE_TEST` will be set to `foo`.

stefano-garzarella replied:
> > IMO we should keep target-svsm and target-user* separated.
> 
> We can separate them, but then we would have to specify a new variable for each one of them. (only `USERINIT_ARGS_TEST` for now) and also `xbuild` should be updated accordingly (to set that value) as it will be possible to have different feature for each usermodule and a single `USERMODULE_ARGS_TEST` would not help
> 
> We should also concatenate all these variables into `--feature` of `xbuild`

Yes, or use 2 different makefiles, one for kernel and one for userspace.

00xc replied:
> A better solution would be to avoid to call back into the makefile, as @00xc suggested in https://github.com/coconut-svsm/svsm/pull/909:
bin/test-kernel.elf is still built from the Makefile for now, since it requires a special command line invocation, and moving that logic into xbuild would introduce too much complexity. Thus, we have xbuild call back into the Makefile to build it

I am actually working on moving everything into xbuild, not sure if it will work but let's see

n-ramacciotti replied:
> I am actually working on moving everything into xbuild, not sure if it will work but let's see

@00xc Great! 

About `xbuild`: https://github.com/coconut-svsm/svsm/pull/926#discussion_r3404368449

00xc replied:
> > I am actually working on moving everything into xbuild, not sure if it will work but let's see
> 
> @00xc Great!

Please take a look at #1127

> About `xbuild`: [#926 (comment)](https://github.com/coconut-svsm/svsm/pull/926#discussion_r3404368449)

Added a comment there

stefano-garzarella commented on user/init/src/main.rs:
```diff
@@ -42,3 +43,10 @@ fn main() -> u32 {
 }
+
+#[cfg(test)]
+fn main() -> u32 {
```
IIRC we discussed this a bit, instead of adding this in user/init/src/main.rs why not creating something different, like user/test_runner/src/main.rs

init.rs should only be an init (to be extended later when we will have an userspace)

n-ramacciotti replied:
In my view, the loop in `test_in_svsm_task` loops over each userbinary compiled for testing mode. Each binary has to have its runner (imported from `userlib`) which will run the tests of that binary only.

Each userbinary will have:
```rust
#![cfg_attr(
    all(test, test_in_svsm),
    no_main,
    feature(custom_test_frameworks),
    test_runner(userlib::testing::svsm_userspace_test_runner),
    reexport_test_harness_main = "userspace_test_main"
)]
```

and will call `userspace_test_main` somewhere (in this case a main function available with #[cfg(test)].

When compiled for testing, `init` will just run the test found in the `init` crate tagged with:
```rust
#[cfg_attr(not(test_in_svsm), ignore = "Can only be run inside guest")]
```

Currently, only the 'init' module is available. However, when other modules become available, 'init' will not interact with them.

I don't understand what the purpose of a `test_runner` crate is in this scenario. @stefano-garzarella could you please elaborate  further.

stefano-garzarella replied:
okay, maybe I expressed myself wrong, but in this case those tests are not really related to init right (also because our init is not doing anythin)?
(I know it's pre-existing)

So maybe better to add a new crate just for testing userspace support of svsm (that maybe we can extend for example to test syscall, etc.)

n-ramacciotti replied:
>  those tests are not really related to init right 

Right, those tests have nothing to do with `init`. Is it possible that `init` will  run related tests? In that case it should follow this structure

> add a new crate just for testing userspace support of svsm

Do you mean a dummy crate that contains only tests (for that crate), which are called by the (userlib) runner when the crate is compiled for testing and it is not used in normal builds?

stefano-garzarella replied:
> Do you mean a dummy crate that contains only tests (for that crate), which are called by the (userlib) runner when the crate is compiled for testing and it is not used in normal builds?

yep, something like that.

n-ramacciotti replied:
Ok, I will move the tests from `init` into the new crate. 

However, it is still possible that the init function will have tests in the future and will have to follow this structure. (I'll drop this for now in its crate.)

stefano-garzarella replied:
> However, it is still possible that the init function will have tests in the future and will have to follow this structure. (I'll drop this for now in its crate.)

yeah, absolutely!

stefano-garzarella reviewed (commented):
I left some comments after a first look. Thanks for this work!

n-ramacciotti committed:
userlib/console: use `$crate` in macro definitions

Every time `println!` is used, it is necessary to import `print!`, as
it is used internally by the macro.

Without it, the following error arises:
```
error: cannot find macro `print` in this scope
  --> user/lib/src/console.rs:44:23
   |
44 |     ($($arg:tt)*) => (print!("{}\n", format_args!($($arg)*)));
   |                       ^^^^^
   |
  ::: user/lib/src/testing.rs:7:5
   |
7  |     println!("running {} user tests", test_cases.len());
   |     ---------------------------------------------------
					in this macro invocation
   |
   = help: have you added the `#[macro_use]` on the module/import?
   = note: this error originates in the macro `println` (in Nightly
	builds, run with -Z macro-backtrace for more info)
help: consider importing this macro through its public re-export
  --> user/lib/src/testing.rs:3:1
   |
3  + use crate::print;
   |
```

The same happens when using `print!` as it requires `console_print`.

Prefix `console_print` and `print!` with `$crate` inside the macro
definitions to ensure that they uses the correct path without requiring
the imports.

Suggested-by: Stefano Garzarella <sgarzare@redhat.com>
Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

n-ramacciotti committed:
kernel/test-in-svsm: Rename test harness main

Change the name of the test harness main and of the test runner to make
it clear that they run kernel tests when testing in svsm.

Add a comment to explain how it works internally.

Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

n-ramacciotti committed:
kernel/test-in-svsm: Move exit from QEMU out of test runner

Currently, the kernel test-in-svsm runner exits directly when all tests
succeed. This prevents the addition of userspace tests.

To introduce userspace test-in-svsm, move the `exit` (from QEMU) call out
of the runner and in `test_in_svsm_task`, which in a subsequent commit
will also be responsible for invoking userspace tests.

Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

n-ramacciotti committed:
user/uapi_tester: Add new crate for testing purposes

Currently, the init process prints a string and then performs some checks
on global mutable data.

With the new test runner, these tests can be moved into a dedicated test
function that can be executed during runtime testing. While moving them,
remove the soft-deprecated macros `addr_of` and `addr_of_mut`.

Add a new crate to test syscalls to keep init clean and move the new test
function there.

Make sure it is compiled only during testing and clippy excludes it for
normal builds.

Add it to the build recipes with the new `test` xbuild type introduced
for kernel test-in-svsm, which can be used also for the userspace.

Signed-off-by: Nicola Ramacciotti <niko.ramak@gmail.com>

luigix25 commented on kernel/src/svsm.rs:
```diff
@@ -623,7 +623,43 @@ pub unsafe fn svsm_main(li: usize) {
+    use svsm::task::{exec_user, wait_for_termination};
+
+    let files = list_dir("/test")?;
+    log::info!("Files: {files:?}");
```
probably this can be removed

n-ramacciotti replied:
Yes, thanks!

luigix25 commented on user/kapi_tester/src/main.rs:
```diff
@@ -0,0 +1,62 @@
+#[cfg(not(test))]
+fn main() -> u32 {
+    0
+}
```
can we drop this main?

luigix25 reviewed (commented):
just nits, rest LGTM!

stefano-garzarella commented on kernel/src/svsm.rs:
```diff
@@ -629,9 +629,36 @@ fn test_in_svsm_task(_context: usize) {
+        wait_for_termination(task.clone());
+        let exit_status = task.get_exit_status()?;
+        if exit_status != TaskExitStatus::Exited(0) {
+            return Err(SvsmError::Test);
```
TBH I don't like to add a new `SvsmError::Test` just for test. IMO we should avoid that.

Also I think it's useful to print the TaskExitStatus value we received for debugging purpose.

What about the following options:
1. move `exit(QEMUExitValue::Fail)` in `user_tests_in_svsm()` , use `unwrap()` instead of `?` (the panic handler should already call `exit(QEMUExitValue::Fail)` ) and remove the return value from user_tests_in_svsm() like we do for kernel_tests_in_svsm()
2. return another value already defined like `SvsmError::Fault`

I prefer 1, it's perfectly fine to panic in a test (calling unwrap, etc.) in a case of failure. I also think we don't even need to move `exit(QEMUExitValue::Fail)`, just add an assert on the `exit_status`.

n-ramacciotti replied:
I'll change the code to use option 1 with the assertion

stefano-garzarella commented on kernel/src/svsm.rs:
```diff
@@ -629,9 +629,36 @@ fn test_in_svsm_task(_context: usize) {
+    use svsm::task::TaskExitStatus;
+    use svsm::task::{exec_user, wait_for_termination};
+
+    for file in list_dir("/test")? {
```
if the `/test` is missing, we are going to fail, right?

Is this intentional, or we should treat is like we don't have any test to run?

n-ramacciotti replied:
> we should treat is like we don't have any test to run?

Right, it's better to assume that there are no tests

stefano-garzarella commented on user/kapi_tester/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+[package]
+name = "kapi_tester"
```
`kapi` is confusing as name IMO.
In Linux we use the opposite (e.g. `uapi` to refer to the API exposed by the kernel to the userspace).

n-ramacciotti replied:
I will change it to `uapi`

stefano-garzarella commented on kernel/src/svsm.rs:
```diff
@@ -629,9 +629,36 @@ fn test_in_svsm_task(_context: usize) {
 }
 
+#[cfg(test)]
+fn user_tests_in_svsm() -> Result<(), SvsmError> {
```
What about moving this in `kernel/src/testing.rs` ?

n-ramacciotti replied:
Makes sense!

stefano-garzarella commented on kernel/src/testing.rs:
```diff
@@ -86,9 +91,28 @@ pub fn svsm_test_runner(test_cases: &[&test::TestDescAndFn]) {
+/// value and panics if the binary exits with a code different from zero.
+/// If the directory is not present, it is assumed that there are no tests to run.
+pub fn user_tests_in_svsm() {
+    if let Ok(files) = list_dir("/test") {
```
nit: What about this? (I found it more readable, but not a strong opinion)

```suggestion
    let Ok(files) = list_dir("/test") else {
        log::info!("/test dir not found, skipping user tests in svsm");
        return;
      };

      for file in files {
```

stefano-garzarella commented on Makefile:
```diff
@@ -157,13 +157,14 @@ endif
 	${CARGO} clippy ${CLIPPY_OPTIONS} --package stage1 --target x86_64-unknown-none -- ${CLIPPY_ARGS}
 	${CARGO} clippy ${CLIPPY_OPTIONS} --workspace --tests --exclude svsm -- ${CLIPPY_ARGS}
 	${CARGO} clippy ${CLIPPY_OPTIONS} --package svsm --tests -- ${CLIPPY_ARGS}
+	${CARGO} clippy ${CLIPPY_OPTIONS} --package uapi_tester --tests -- ${CLIPPY_ARGS}
```
Do we really need this?
I mean my doubt if this is already covered by line 165

n-ramacciotti replied:
Oh, right!

stefano-garzarella commented on user/uapi_tester/src/main.rs:
```diff
@@ -0,0 +1,57 @@
+            write(&mut *addr_of_mut!(SOME_BSS_DATA), 0xaa);
+            check(&*addr_of!(SOME_DATA), 0xccu64);
+            check(&*addr_of!(SOME_RO_DATA), 0xeeu64);
+            check(&*addr_of!(SOME_BSS_DATA), 0xaa);
```
nit: not strong opinion, and also not sure if it makes the code better, but should we take this opportunity to replace `addr_of` and `addr_of_mut` ?

See https://doc.rust-lang.org/core/ptr/macro.addr_of.html
> addr_of!(expr) is equivalent to &raw const expr. The macro is soft-deprecated; use &raw const instead.

stefano-garzarella commented on user/uapi_tester/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+
+[lib]
+test = true
+doctest = true
```
Not sure if it's a good idea. IIRC `doctest`s use std, but maybe I'm missing something and there is a reason to set this to true.

n-ramacciotti replied:
From an old conversation in this PR (https://github.com/coconut-svsm/svsm/pull/926#discussion_r2640448089): By default these values are already set to true, so this section can be removed. It is here for clarity and for consistency with `kernel/Cargo.toml`, which contains the same section.

stefano-garzarella replied:
@n-ramacciotti I see, if we have in `kernel/Cargo.toml` let's leave them, we can figure out later if there are issues.

stefano-garzarella reviewed (commented):
LGTM, mostly nit comments plus the makefile rule which I'm not sure.
